<img src="https://github.com/hernancontigiani/ceia_memorias_especializacion/raw/master/Figures/logoFIUBA.jpg" width="500" align="center">

# Desafio 1 PLN CEIA Ojeda, Juan

### Consigna del desafío 1

**1**. Vectorizar documentos. Tomar 5 documentos al azar y medir similaridad con el resto de los documentos.
Estudiar los 5 documentos más similares de cada uno analizar si tiene sentido
la similaridad según el contenido del texto y la etiqueta de clasificación.

**2**. Entrenar modelos de clasificación Naïve Bayes para maximizar el desempeño de clasificación
(f1-score macro) en el conjunto de datos de test. Considerar cambiar parámteros
de instanciación del vectorizador y los modelos y probar modelos de Naïve Bayes Multinomial
y ComplementNB.

**3**. Transponer la matriz documento-término. De esa manera se obtiene una matriz
término-documento que puede ser interpretada como una colección de vectorización de palabras.
Estudiar ahora similaridad entre palabras tomando 5 palabras y estudiando sus 5 más similares. **La elección de palabras no debe ser al azar para evitar la aparición de términos poco interpretables, elegirlas "manualmente"**.


### Vectorización de texto y modelo de clasificación Naïve Bayes con el dataset 20 newsgroups

In [1]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.naive_bayes import MultinomialNB, ComplementNB
from sklearn.metrics import f1_score
import pandas as pd

# 20newsgroups por ser un dataset clásico de NLP ya viene incluido y formateado
# en sklearn
from sklearn.datasets import fetch_20newsgroups
import numpy as np

doc1 = 1010
doc2 = 2020
doc3 = 3030
doc4 = 4040
doc5 = 5050
docs = [doc1, doc2, doc3, doc4, doc5]

## Carga de datos

In [2]:
# cargamos los datos (ya separados de forma predeterminada en train y test)
newsgroups_train = fetch_20newsgroups(subset='train', remove=('headers', 'footers', 'quotes'))
newsgroups_test = fetch_20newsgroups(subset='test', remove=('headers', 'footers', 'quotes'))

## Similaridad de documentos

In [8]:
tfidfvect = TfidfVectorizer()
X_train = tfidfvect.fit_transform(newsgroups_train.data)
y_train = newsgroups_train.target
cossim = cosine_similarity(X_train[docs], X_train)

In [9]:
# Buscamos los documentos más similares a los seleccionados
for i, doc_idx in enumerate(docs):
    similares = np.argsort(cossim[i])[::-1][1:6]
    print(f"Documento base: {doc_idx} - Clase: {newsgroups_train.target_names[y_train[doc_idx]]}")
    for sim_idx in similares:
        print(f"   Similar: {sim_idx} - Clase: {y_train[sim_idx]} - Target: {newsgroups_train.target_names[y_train[sim_idx]]}")


Documento base: 1010 - Clase: rec.autos
   Similar: 1025 - Clase: 7 - Target: rec.autos
   Similar: 1144 - Clase: 7 - Target: rec.autos
   Similar: 6773 - Clase: 7 - Target: rec.autos
   Similar: 9623 - Clase: 17 - Target: talk.politics.mideast
   Similar: 6437 - Clase: 17 - Target: talk.politics.mideast
Documento base: 2020 - Clase: sci.med
   Similar: 3306 - Clase: 13 - Target: sci.med
   Similar: 6894 - Clase: 16 - Target: talk.politics.guns
   Similar: 5826 - Clase: 15 - Target: soc.religion.christian
   Similar: 9623 - Clase: 17 - Target: talk.politics.mideast
   Similar: 4988 - Clase: 13 - Target: sci.med
Documento base: 3030 - Clase: soc.religion.christian
   Similar: 1619 - Clase: 15 - Target: soc.religion.christian
   Similar: 5213 - Clase: 15 - Target: soc.religion.christian
   Similar: 913 - Clase: 0 - Target: alt.atheism
   Similar: 10836 - Clase: 0 - Target: alt.atheism
   Similar: 523 - Clase: 15 - Target: soc.religion.christian
Documento base: 4040 - Clase: rec.sport.bas

---

## 1.

Se resume a continuación, para el documento 1010 (`doc1`), el análisis de similitud con otros documentos:

| Índice | Clase               | ¿Relacionado con autos? | ¿Similar al 1010? | Comentario breve                                      |
|--------|---------------------|--------------------------|-------------------|--------------------------------------------------------|
| 1010   | rec.autos           | Parcial               | —                 | Historia con una Toyota, pero no trata de autos       |
| 1025   | rec.autos           | Sí                   | Sí              | Problemas mecánicos, temática más técnica             |
| 1144   | rec.autos           | Sí                   | Parcial         | Coincidencia léxica más que semántica                 |
| 6773   | rec.autos           | Sí                   | Sí              | Falla mecánica en la ruta                             |
| 9623   | talk.politics.mideast | No                 | No              | Política, violencia, palabras engañosas               |
| 6437   | talk.politics.mideast | No                 | No              | Economía y conflictos sociales                        |

En conclusión, el documento 1010 no está fuertemente relacionado con su clase asignada (`rec.autos`), aunque comparte ciertas palabras clave. Documentos como el 1025, 1144 y 6773 sí están bien clasificados dentro de la temática automotor. Por otro lado, los documentos 9623 y 6437 pertenecen a una categoría completamente diferente, pero podrían haber aparecido como similares debido al tono violento o conflictivo del relato original.

---

Se resume a continuación, para el documento 2020 (`doc2`), el análisis de similitud con otros documentos:

| Índice | Clase                 | ¿Relacionado con medicina? | ¿Similar al 2020? | Comentario breve                                                                  |
|--------|-----------------------|-----------------------------|-------------------|-----------------------------------------------------------------------------------|
| 2020   | sci.med               | Sí                     | —                 | Se menciona un problema médico en una experiencia personal                        |
| 3306   | sci.med               | Sí                          | Sí                | Discusión sobre efectos secundarios y tratamientos                                |
| 6894   | talk.politics.guns    | No                          | No                | Debate político sobre armas, sin relación médica                                  |
| 5826   | soc.religion.christian| No                          | No                | Enfoque religioso, sin vínculo con la medicina                                    |
| 9623   | talk.politics.mideast | No                          | No                | Opinión política sobre conflictos, vocabulario puede coincidir superficialmente   |
| 4988   | sci.med               | Sí                          | Sí                | Consulta médica y referencias a síntomas                                          |

En conclusión, el documento 2020 está correctamente clasificado dentro de la categoría `sci.med`, aunque su contenido se mezcla con una anécdota personal. Los documentos 3306 y 4988 también pertenecen a `sci.med` y **presentan coincidencias temáticas claras. Por otro lado, los documentos 6894, 5826 y 9623 pertenecen a categorías completamente distintas (política, religión) y su aparición como similares puede deberse a la coincidencia de términos como “problema”, “riesgo”, “personas” o “conflicto”, comunes tanto en textos médicos como en debates sociales.

---

Se resume a continuación, para el documento 3030 (`doc3`), el análisis de similitud con otros documentos:

| Índice | Clase                 | ¿Relacionado con religión cristiana? | ¿Similar al 3030? | Comentario breve                                                                 |
|--------|-----------------------|--------------------------------------|-------------------|----------------------------------------------------------------------------------|
| 3030   | soc.religion.christian| Sí                                   | —                 | Discurso centrado en la visión cristiana sobre el universo                      |
| 1619   | soc.religion.christian| Sí                                   | Sí                | Debate sobre fe y creencias religiosas                                           |
| 5213   | soc.religion.christian| Sí                                   | Sí                | Reflexión religiosa con referencias bíblicas                                     |
| 913    | alt.atheism           | No                                   | Parcial           | Contraposición de ideas con perspectiva atea                                     |
| 10836  | alt.atheism           | No                                   | Parcial           | Postura crítica hacia la religión, posible cruce temático                       |
| 523    | soc.religion.christian| Sí                                   | Sí                | Referencias directas a la fe cristiana                                           |

En conclusión, el documento 3030 está correctamente clasificado dentro de la categoría `soc.religion.christian`, ya que aborda temas religiosos desde una perspectiva cristiana. Los documentos 1619, 5213 y 523 también comparten esta orientación y son coherentes tanto temática como léxicamente. Los documentos 913 y 10836 pertenecen a la categoría `alt.atheism`, y aunque tienen una postura opuesta, pueden haber aparecido como similares debido a la coincidencia de términos como “Dios”, “Biblia”, “fe” o “iglesia”, comunes a ambos discursos aunque desde enfoques diferentes.

---

Se resume a continuación, para el documento 4040 (`doc4`), el análisis de similitud con otros documentos:

| Índice | Clase               | ¿Relacionado con béisbol? | ¿Similar al 4040? | Comentario breve                                                                     |
|--------|---------------------|----------------------------|-------------------|----------------------------------------------------------------------------------------|
| 4040   | rec.sport.baseball  | Sí                    | —                 | Comentario breve, pero con firma común y contexto deportivo implícito                 |
| 3981   | rec.sport.baseball  | Sí                         | Sí                | Referencia a un jugador, firma repetida, claramente vinculado al béisbol              |
| 10125  | talk.politics.misc  | No                         | No                | Opinión sobre expresidentes, sin relación temática con deporte                        |
| 9961   | rec.sport.baseball  | Sí                         | Sí                | Mención de comentaristas, transmisiones y equipos; tema vinculado al béisbol          |
| 3354   | rec.sport.baseball  | Sí                         | Sí                | Comentario sobre un juego, jugadores y estrategia; muy relacionado                    |
| 2413   | rec.sport.baseball  | Parcial                    | Parcial           | Texto muy corto y ambiguo, pero contexto indica relación con deporte por proximidad   |

En conclusión, el documento 4040 pertenece correctamente a la categoría `rec.sport.baseball`, aunque su contenido por sí solo es breve y no completamente informativo. Sin embargo, comparte firma y estilo con otros documentos del mismo autor, lo que puede haber ayudado al modelo a agruparlo correctamente. Los documentos 3981, 9961 y 3354 presentan una fuerte coherencia temática, mientras que el documento 2413 tiene un texto escueto, pero parece formar parte del mismo hilo. Por otro lado, el documento 10125 pertenece a una categoría completamente distinta y no tiene ninguna relación temática con el deporte, por lo que su aparición como similar probablemente sea un falso positivo.

---

Se resume a continuación, para el documento 5050 (`doc5`), el análisis de similitud con otros documentos:

| Índice | Clase                 | ¿Relacionado con política en Medio Oriente? | ¿Similar al 5050? | Comentario breve                                                                 |
|--------|-----------------------|---------------------------------------------|-------------------|----------------------------------------------------------------------------------|
| 5050   | talk.politics.mideast | No                                          | —                 | Insulto directo, sin contenido temático                                          |
| 9699   | talk.politics.mideast | No                                          | Sí                | Mismo tipo de mensaje ofensivo, sin contenido político                           |
| 9983   | talk.politics.mideast | No                                          | Sí                | Mensaje agresivo, tono similar, sin relación con política                        |
| 2538   | talk.politics.mideast | No                                          | Sí                | Ofensa directa y personalizada, idéntico estilo retórico                         |
| 1443   | talk.politics.misc    | Parcial                                     | Parcial           | Texto político-legal, pero el término "loser" aparece repetidamente              |
| 6185   | talk.politics.mideast | No                                          | Sí                | Mismo patrón de insulto, pero sin contexto temático definido                     |

En conclusión, el documento 5050 y sus similares forman parte de una serie de mensajes con contenido ofensivo o provocador, sin relación real con política ni con Medio Oriente. A pesar de estar todos etiquetados como `talk.politics.mideast`, la similitud entre ellos se debe al tono y uso repetido del término “loser”, no al contenido. Este ejemplo evidencia una debilidad del modelo, que tiende a agrupar documentos con lenguaje similar aun cuando no hay coherencia temática con la clase asignada.

---

### 2. Modelos de clasificación Naïve Bayes

In [15]:
def comparar_modelos_nb(train_data, test_data, y_train, y_test):
    resultados = []
    
    # Distintas configuraciones del vectorizador
    vectorizers = [
        ("default", TfidfVectorizer()),
        ("stopwords+bigrams", TfidfVectorizer(
                                                stop_words='english',
                                                ngram_range=(1, 2)
                                                )),
        ("filtered+sublinear", TfidfVectorizer(
                                                stop_words='english',
                                                max_df=0.75, min_df=5,
                                                sublinear_tf=True
                                                )),
    ]
    
    # Distintas configuraciones de modelos
    modelos = [
        ("MultinomialNB α=1.0", MultinomialNB(alpha=1.0)),
        ("MultinomialNB α=0.5", MultinomialNB(alpha=0.5)),
        ("ComplementNB α=1.0", ComplementNB(alpha=1.0)),
        ("ComplementNB α=0.5", ComplementNB(alpha=0.5)),
    ]
    
    for nombre_vect, vect in vectorizers:
        X_train_vect = vect.fit_transform(train_data)
        X_test_vect = vect.transform(test_data)
        
        for nombre_modelo, modelo in modelos:
            modelo.fit(X_train_vect, y_train)
            y_pred = modelo.predict(X_test_vect)
            f1 = f1_score(y_test, y_pred, average='macro')
            
            resultados.append({
                "Vectorizador": nombre_vect,
                "Modelo": nombre_modelo,
                "F1 macro": f1
            })
    
    return pd.DataFrame(resultados).sort_values(by="F1 macro", ascending=False)


resultados_nb = comparar_modelos_nb(newsgroups_train.data,
                                    newsgroups_test.data,
                                    newsgroups_train.target,
                                    newsgroups_test.target)
resultados_nb


,Vectorizador,Modelo,F1 macro
7,stopwords+bigrams,ComplementNB α=0.5,0.710536
6,stopwords+bigrams,ComplementNB α=1.0,0.703928
3,default,ComplementNB α=0.5,0.696116
2,default,ComplementNB α=1.0,0.692953
10,filtered+sublinear,ComplementNB α=1.0,0.682009
11,filtered+sublinear,ComplementNB α=0.5,0.681246
9,filtered+sublinear,MultinomialNB α=0.5,0.658301
5,stopwords+bigrams,MultinomialNB α=0.5,0.654455
8,filtered+sublinear,MultinomialNB α=1.0,0.646120
4,stopwords+bigrams,MultinomialNB α=1.0,0.642496


## Conclusión 

El mejor desempeño se obtuvo con el modelo ComplementNB con alfa igual a 0.5, usando un vectorizador que elimina stopwords y considera bigramas. Esta combinación alcanzó un F1 macro de 0.7105.

Esto sugiere que ComplementNB se adapta mejor a datos desbalanceados, y que el uso de bigramas junto con la eliminación de palabras vacías mejora la representación de los documentos. En contraste, el modelo MultinomialNB sin ajustes tuvo el peor desempeño, lo que evidencia la importancia de probar distintas configuraciones para optimizar resultados.

---


In [18]:
# Vectorizador con parámetros ajustados para análisis de términos
vectorizador = TfidfVectorizer(stop_words='english', max_df=0.75, min_df=5, sublinear_tf=True)
X = vectorizador.fit_transform(newsgroups_train.data)

# Transponer la matriz para pasar de documento-término a término-documento
X_td = X.T  # n_terms x n_documents

# Vocabulario en orden del índice
terms = np.array(vectorizador.get_feature_names_out())

# Se eligen 5 palabras manualmente, relevantes para el dataset
palabras_objetivo = ['god', 'car', 'windows', 'israel', 'space']

# Se calculan las 5 palabras más similares para cada una
for palabra in palabras_objetivo:
    if palabra in terms:
        idx = np.where(terms == palabra)[0][0]
        sims = cosine_similarity(X_td[idx], X_td).flatten()
        top5_idx = sims.argsort()[::-1][1:6]
        similares = terms[top5_idx]
        print(f"\nPalabra: {palabra}\nPalabras más similares: {', '.join(similares)}")



Palabra: god
Palabras más similares: jesus, faith, christ, bible, lord

Palabra: car
Palabras más similares: cars, dealer, civic, owner, engine

Palabra: windows
Palabras más similares: dos, ms, files, file, microsoft

Palabra: israel
Palabras más similares: israeli, arab, arabs, lebanon, syrian

Palabra: space
Palabras más similares: nasa, shuttle, launch, astronaut, sci


## Conclusión

Se seleccionaron manualmente cinco palabras relevantes del corpus: god, car, windows, israel y space. Para cada una, se calcularon las cinco palabras más similares usando la matriz término-documento y la similitud del coseno.

Los resultados muestran agrupamientos semánticos esperables. Por ejemplo, "god" se asocia con términos religiosos como "jesus" y "bible", mientras que "car" se relaciona con "engine", "dealer" y "civic". También se observan asociaciones válidas en contextos tecnológicos ("windows" con "microsoft" y "files"), políticos ("israel" con "arab" y "lebanon") y científicos ("space" con "nasa" y "shuttle").

Esto confirma que la representación vectorial captura adecuadamente el contexto temático de las palabras.
